File: frontend/mobile/lib/services/api_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

/**
 * Standard API Client for BizFlow Mobile
 * - Auto-attaches JWT Bearer token
 * - Handles 401 errors with token clearing
 */

class ApiClient {
  static const String _baseUrl = 'http://10.0.2.2:8080/api/v1'; // Default for Android Emulator

  // Helper to get headers with Auth
  static Future<Map<String, String>> _getHeaders() async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('access_token');
    
    return {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
      if (token != null) 'Authorization': 'Bearer $token',
    };
  }

  // Handle Response & 401
  static http.Response _handleResponse(http.Response response) {
    if (response.statusCode == 401) {
      // Clear token on 401
      SharedPreferences.getInstance().then((prefs) {
        prefs.remove('access_token');
        prefs.remove('user_data');
      });
      throw Exception('Session expired. Please login again.');
    }
    return response;
  }

  // GET
  static Future<http.Response> get(String endpoint) async {
    final url = Uri.parse('$_baseUrl$endpoint');
    final response = await http.get(url, headers: await _getHeaders());
    return _handleResponse(response);
  }

  // POST
  static Future<http.Response> post(String endpoint, dynamic body) async {
    final url = Uri.parse('$_baseUrl$endpoint');
    final response = await http.post(
      url, 
      headers: await _getHeaders(),
      body: jsonEncode(body),
    );
    return _handleResponse(response);
  }

  // PUT
  static Future<http.Response> put(String endpoint, dynamic body) async {
    final url = Uri.parse('$_baseUrl$endpoint');
    final response = await http.put(
      url, 
      headers: await _getHeaders(),
      body: jsonEncode(body),
    );
    return _handleResponse(response);
  }

  // DELETE
  static Future<http.Response> delete(String endpoint) async {
    final url = Uri.parse('$_baseUrl$endpoint');
    final response = await http.delete(url, headers: await _getHeaders());
    return _handleResponse(response);
  }
}

File: frontend/web/lib/apiClient.ts
import axios, { AxiosInstance, InternalAxiosRequestConfig } from 'axios';

/**
 * Standard API Client for BizFlow Web
 * - Auto-attaches JWT Bearer token
 * - Handles 401/403 errors with auto-logout
 * - Uses environment variable for baseURL
 */

const baseURL = (process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8080').replace(/\/api\/v1\/?$/, '') + '/api/v1';

const apiClient: AxiosInstance = axios.create({
  baseURL,
  headers: {
    'Content-Type': 'application/json',
  },
  timeout: 30000,
});

// Request Interceptor: Attach Token
apiClient.interceptors.request.use(
  (config: InternalAxiosRequestConfig) => {
    if (typeof window !== 'undefined') {
      const token = localStorage.getItem('access_token');
      if (token) {
        config.headers.Authorization = `Bearer ${token}`;
      }
    }
    return config;
  },
  (error) => Promise.reject(error)
);

let isRedirecting = false;

// Response Interceptor: Handle Auth Errors
apiClient.interceptors.response.use(
  (response) => response,
  async (error) => {
    // Only handle auth errors if we have a response
    if (error.response?.status === 401) {
      // 401 means token is invalid or missing - this is a real auth failure
      if (typeof window !== 'undefined' && !window.location.pathname.includes('/login')) {
        if (isRedirecting) return Promise.reject(error);

        isRedirecting = true;
        console.warn('Token không hợp lệ hoặc đã hết hạn. Đang đăng xuất...');

        // Dispatch event for UI to show toast if needed
        window.dispatchEvent(new CustomEvent('bizflow:auth-error', {
          detail: { message: 'Phiên đăng nhập đã hết hạn. Vui lòng đăng nhập lại.' }
        }));

        // Access store and logout
        try {
          const { useAuthStore } = await import('@/store/useAuthStore');
          useAuthStore.getState().logout();
        } catch (e) {
          localStorage.removeItem('access_token');
          localStorage.removeItem('bizflow-auth');
        }

        // Redirect to login
        setTimeout(() => {
          isRedirecting = false;
          window.location.href = '/login?error=session_expired';
        }, 500);
      }
    }
    else if (error.response?.status === 403) {
      // 403 means user is authenticated but not authorized for this resource
      // This is NOT a session expiry - just show error, don't logout
      console.warn('Không có quyền truy cập tài nguyên này:', error.config.url);

      // Dispatch event for UI to show permission error
      window.dispatchEvent(new CustomEvent('bizflow:permission-error', {
        detail: { message: 'Bạn không có quyền thực hiện thao tác này.' }
      }));
    } else if (!error.response) {
      // Network error - no response from server
      console.error('Lỗi kết nối mạng:', error.message);

      if (typeof window !== 'undefined') {
        window.dispatchEvent(new CustomEvent('bizflow:network-error', {
          detail: { message: 'Không thể kết nối đến máy chủ. Vui lòng kiểm tra kết nối mạng.' }
        }));
      }
    }

    return Promise.reject(error);
  }
);

export default apiClient;
